docs/rfcs/20220209_limit_connections_rate.md

```diff
@@ -0,0 +1,134 @@
+- Feature Name: Ability to limit connection creation rate per shard
+- Status: in-progress
+- Start Date: 2022-02-09
+- Authors: Vadim Plakhtinskiy
+- Issue: 
+
+# Executive Summary
+
+- Set a limit on the rate with which the shard accepts new connections
+- Set a limit on connection creation rate per IP
+
+
+Similar to KIP-612, adapted to Redpanda's thread per core model
+
+## What is being proposed
+
+New configuration properties (For start we choose global config for connection rate):
```

Comment:
What's the time window? Can we please make the time window configurable as well? Also, it is useful to provide some configuration examples and explain how the algorithm will proceed since not everyone is familiar with the internals of token bucket algorithms.

Reply:
This settings it similar to` max.connection.creation.rate` in kafka
I think use do not want to set window length (kafka do not provide it).

Maybe we should add it like new option in future if we really need it

Reply:
It does provide it through https://cwiki.apache.org/confluence/display/KAFKA/KIP-124+-+Request+rate+quotas. 

From KIP-612:
> Connection rate throttling behavior
The connection creation rate limits will be applied to the same quota window configuration (quota.window.size.seconds with 1 second default) as existing produce/fetch quotas and request rate quota ([KIP-124](https://cwiki.apache.org/confluence/display/KAFKA/KIP-124+-+Request+rate+quotas)). Since limit on connection creation rate on the broker is also a type of quota, this approach will keep it consistent with the existing quota implementations on the broker. If connection creation rate on the broker exceeds the broker-wide limit, the broker will delay accepting a new connection by an amount of time that brings the rate within the limit. If a listener-specific limit is specified, and the connection rate on that listener exceeds the limit, the broker will delay accepting new connection on that listener by an amount of time that brokers the rate of the listener within the listener limit. The maximum delay applied will be the quota window size (which also means that the minimum connection rate limit is effectively 1 connection creation / second).

Reply:
We have another rates:
`target_quota_byte_rate`
`raft_learner_recovery_rate`

I think adding rate window should be separate task.
We use ss::lowers_clock and `This is a monotonic clock with a granularity of 10 ms`, can it be a problem?

Maybe @jcsp or @dotnwat can help with this